### DISCUSSION TITLE
Why microservice?

### DISCUSSION THREAD

**Discussion by LLLLimbo** [edgexfoundry/community#54] [General]
is there an article on why EdgeX chose to use microservice architecture?

**Comment by jpwhitemn** [comment]
@LLLLimbo - there are several reasons why we choose a microservice architecture (some are listed in our documentation here:  https://docs.edgexfoundry.org/2.3/general/Definitions/#micro-service).  Essentially, we wanted the ability to replace blocks of functionality quickly (with best of breed solutions that may include services that come from outside the project).  We wanted the ability to provide the services in any language or technology of your choice.  We also believed that your use case would require different capability so flexibility to add or remove services was paramount (like device services for the specific types of sensors/devices you needed to connect to or an app service for the particular cloud system you needed to communicate with).  Lastly, because of the resource constraints often found at the edge, we believed it was necessary to be able to distributed EdgeX (via micro service distribution) across the compute, network, etc. resources available at the edge (or even beyond the edge).  Hope this helps.

**Reply by LLLLimbo** [reply]
Thank you for your answer, it was very helpful!
